File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/ui/fragments/MeteoriteDetailFragment.java
package com.antonio.samir.meteoritelandingsspots.ui.fragments;

import android.arch.lifecycle.LiveData;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.antonio.samir.meteoritelandingsspots.R;
import com.antonio.samir.meteoritelandingsspots.model.Meteorite;
import com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteRepositoryFactory;
import com.antonio.samir.meteoritelandingsspots.service.repository.database.MeteoriteDao;
import com.antonio.samir.meteoritelandingsspots.util.analytics.AnalyticsUtil;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import org.apache.commons.lang3.StringUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MeteoriteDetailFragment extends Fragment implements OnMapReadyCallback {

    public static final String METEORITE = "METEORITE";

    @BindView(R.id.title)
    TextView title;
    @BindView(R.id.location)
    TextView location;
    @BindView(R.id.year)
    TextView year;
    @BindView(R.id.mass)
    TextView mass;
    @BindView(R.id.recclass)
    TextView recclass;

    @BindView(R.id.year_label)
    TextView yearLabel;
    @BindView(R.id.mass_label)
    TextView massLabel;
    @BindView(R.id.recclass_label)
    TextView recclassLabel;

    private String meteoriteId;

    private GoogleMap mMap;
    private LiveData<Meteorite> mMeteoriteLiveData;
    private boolean isUiDone = false;

    /**
     * Create a MeteoriteDetailFragment to show the meteorite param
     * @param meteorite
     * @return
     */
    public static MeteoriteDetailFragment newInstance(final String meteorite) {
        final MeteoriteDetailFragment fragment = new MeteoriteDetailFragment();
        final Bundle args = new Bundle();
        args.putString(METEORITE, meteorite);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            //Recovering the meteorite to work on it
            meteoriteId = getArguments().getString(METEORITE);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        final View view = inflater.inflate(R.layout.fragment_meteorite_detail, container, false);

        ButterKnife.bind(this, view);

        ((SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.map)).getMapAsync(this);

        final AppCompatActivity activity = (AppCompatActivity) getActivity();

        final Toolbar toolbarView = view.findViewById(R.id.toolbar);

        if (null != toolbarView) {
            activity.setSupportActionBar(toolbarView);

            activity.getSupportActionBar().setDisplayShowTitleEnabled(false);
            activity.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        setMeteorite(meteoriteId);
    }

    @Override
    public void onMapReady(GoogleMap map) {
        this.mMap = map;
    }

    public void setupMap(String meteoriteName, double lat,double  log) {
        mMap.clear();

        final LatLng latLng = new LatLng(lat, log);

        mMap.addMarker(new MarkerOptions().position(
                latLng).title(meteoriteName));

        mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 30));

        // Zoom in, animating the camera.
        mMap.animateCamera(CameraUpdateFactory.zoomTo(10), 2000, null);
    }

    public void setMeteorite(final String meteoriteId) {
        this.meteoriteId = meteoriteId;

        final MeteoriteDao meteoriteDao = MeteoriteRepositoryFactory.getMeteoriteDao(getContext());

        mMeteoriteLiveData = meteoriteDao.getMeteoriteById(meteoriteId);

        mMeteoriteLiveData.observe(this, this::initView);

    }

    private void initView(final Meteorite meteorite) {
        setLocationText(meteorite.getAddress(), this.location);

        if (!isUiDone) {
            isUiDone = true;
            final String meteoriteName = meteorite.getName();
            setText(null, this.title, meteoriteName);

            final String yearString = meteorite.getYearString();
            setText(yearLabel, this.year, yearString);

            final String recclass = meteorite.getRecclass();
            setText(recclassLabel, this.recclass, recclass);

            final String mass = meteorite.getMass();
            setText(massLabel, this.mass, mass);

            if (mMap != null) {
                final Double lat = Double.valueOf(meteorite.getReclat());
                final Double log = Double.valueOf(meteorite.getReclong());
                setupMap(meteoriteName, lat, log);
            }

            AnalyticsUtil.logEvent("Detail", String.format("%s detail", meteoriteName));
        }
    }

    public void setLocationText(final String address, final TextView text) {
        if (StringUtils.isNotEmpty(address)) {
            setText(null, text, address);
            text.setVisibility(View.VISIBLE);
            mMeteoriteLiveData.removeObservers(this);
        } else {
            text.setVisibility(View.GONE);
        }
    }

    private void setText(TextView textFieldLabel, final TextView textField, final String text) {
        if (StringUtils.isEmpty(text)) {
            if (textFieldLabel != null) {
                textFieldLabel.setVisibility(View.GONE);
            }
            textField.setVisibility(View.GONE);
        } else {
            textField.setText(text);
            textField.setContentDescription(text);
        }
    }

}

File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/ui/fragments/MeteoriteDetailFragment.kt
package com.antonio.samir.meteoritelandingsspots.ui.fragments

import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.antonio.samir.meteoritelandingsspots.R
import com.antonio.samir.meteoritelandingsspots.model.Meteorite
import com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteRepositoryFactory
import com.antonio.samir.meteoritelandingsspots.util.analytics.AnalyticsUtil
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import kotlinx.android.synthetic.main.meteorite_detail.*
import kotlinx.android.synthetic.main.meteorite_detail_grid.*
import org.apache.commons.lang3.StringUtils

class MeteoriteDetailFragment : androidx.fragment.app.Fragment(), OnMapReadyCallback {

    private var meteoriteId: String? = null

    private var mMap: GoogleMap? = null
    private var mMeteoriteLiveData: LiveData<Meteorite>? = null
    private var isUiDone = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            //Recovering the meteorite to work on it
            meteoriteId = arguments!!.getString(METEORITE)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.fragment_meteorite_detail, container, false)

        (childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment).getMapAsync(this)

        val activity = activity as AppCompatActivity?

        val toolbarView = view.findViewById<Toolbar>(R.id.toolbar)

        if (null != toolbarView) {
            activity!!.setSupportActionBar(toolbarView)
            activity.supportActionBar!!.setDisplayShowTitleEnabled(false)
            activity.supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }

        return view
    }

    override fun onStart() {
        super.onStart()
        setMeteorite(meteoriteId)
    }

    override fun onMapReady(map: GoogleMap) {
        this.mMap = map
    }

    fun setupMap(meteoriteName: String, lat: Double, log: Double) {
        mMap?.clear()

        val latLng = LatLng(lat, log)

        mMap?.addMarker(MarkerOptions().position(
                latLng).title(meteoriteName))

        mMap?.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 30f))

        // Zoom in, animating the camera.
        mMap?.animateCamera(CameraUpdateFactory.zoomTo(10f), 2000, null)
    }

    fun setMeteorite(meteoriteId: String?) {
        this.meteoriteId = meteoriteId

        val meteoriteDao = MeteoriteRepositoryFactory.getMeteoriteDao(context!!)

        mMeteoriteLiveData = meteoriteId?.let { meteoriteDao.getMeteoriteById(it) }

        val observer = Observer<Meteorite> {t: Meteorite? -> t?.let { this.initView(it) } }

        mMeteoriteLiveData!!.observe(this, observer)

    }

    private fun initView(meteorite: Meteorite) {
        setLocationText(meteorite.address, view?.findViewById(R.id.location))

        if (!isUiDone) {
            isUiDone = true
            val meteoriteName = meteorite.name
            setText(null, this.title, meteoriteName)

            val yearString = meteorite.yearString
            setText(year_label, this.year, yearString)

            val recclass = meteorite.recclass
            setText(recclass_label, this.recclass, recclass)

            val mass = meteorite.mass
            setText(mass_label, this.mass, mass)

            if (mMap != null) {
                val lat = java.lang.Double.valueOf(meteorite.reclat)
                val log = java.lang.Double.valueOf(meteorite.reclong)
                setupMap(meteoriteName, lat!!, log!!)
            }

            AnalyticsUtil.logEvent("Detail", String.format("%s detail", meteoriteName))
        }
    }

    fun setLocationText(address: String, text: TextView?) {
        if (StringUtils.isNotEmpty(address)) {
            setText(null, text, address)
            text!!.visibility = View.VISIBLE
            mMeteoriteLiveData!!.removeObservers(this)
        } else {
            text!!.visibility = View.GONE
        }
    }

    private fun setText(textFieldLabel: TextView?, textField: TextView?, text: String?) {
        if (StringUtils.isEmpty(text)) {
            if (textFieldLabel != null) {
                textFieldLabel.visibility = View.GONE
            }
            textField!!.visibility = View.GONE
        } else {
            textField!!.text = text
            textField.contentDescription = text
        }
    }

    companion object {

        val METEORITE = "METEORITE"

        /**
         * Create a MeteoriteDetailFragment to show the meteorite param
         * @param meteorite
         * @return
         */
        fun newInstance(meteorite: String): MeteoriteDetailFragment {
            val fragment = MeteoriteDetailFragment()
            val args = Bundle()
            args.putString(METEORITE, meteorite)
            fragment.arguments = args
            return fragment
        }
    }

}
